### PR TITLE
RFC: Add a new config option for configuring enqueue limit of state machin…

### PR DESCRIPTION
**Pull request by mitake**
…e command apply

This commit adds a new parameter to consul agent: enqueue_limit. This
option is used as an enqueue limit of state machine's command
execution. The type of the parameter is time.Duration. If 0 is
specified, timeout will never happen.

Example of usage:
{
     "enqueue_limit": "45s"
}

Supplying such a json via -config-file means the parameter will be
configured 45 seconds instead of the default (30 seconds).

The motivation of this change is handling large number of requests. In the past we faced errors of timeout actually. But if consul clients need timeout for their requests, they can do it by themselves. I thought having a fixed timeout is not so reasonable for large scale deployments.

### CODE REVIEW

**Comment by mitake** [comment]
I found this PR didn't pass the CI. I fixed it and now it can pass.

**Comment by i0rek** [comment]
Thank you for your work! Making `enqueue_limit` configurable seems reasonable to me. Would you be willing to update your PR?

**Comment by mitake** [comment]
@i0rek thanks for your reply! Yeah let me rebase on the latest master later.

**Comment by mitake** [comment]
@i0rek updated based on the latest master branch, could you take a look?

**Comment by mitake** [comment]
@i0rek I found the test case (?) `CASE cfg-resolver-dc-failover-gateways-remote, envoy 1.8.0: Going to remove envoy_verify-primary_1` failed in `test-integrations`. I tried to reproduce it on my local env but the test case seems to be finished successfully. In addition, other test cases like `CASE cfg-resolver-dc-failover-gateways-remote, envoy 1.8.0: Going to remove envoy_verify-primary_1` failed on my local env. Are these test cases freaky?

File: agent/config/config.go [hunk 1]
```diff
@@ -312,6 +312,8 @@ type Config struct {
 	SyncCoordinateRateTarget   *float64 `json:"sync_coordinate_rate_target,omitempty" hcl:"sync_coordinate_rate_target" mapstructure:"sync_coordinate_rate_target"`
 	Version                    *string  `json:"version,omitempty" hcl:"version" mapstructure:"version"`
 	VersionPrerelease          *string  `json:"version_prerelease,omitempty" hcl:"version_prerelease" mapstructure:"version_prerelease"`
+
+	EnqueueLimit *string `json:"enqueue_limit,omitempty" hcl:"enqueue_limit" mapstructure:"enqueue_limit"`
```
Comment (i0rek) [inline agent/config/config.go]: Please move that up a little, it ended up under `// non-user configurable values` which is not correct. Below `Watches` seems like a good place, would be similar to the position in `builder.go`.

File: agent/config/runtime.go [hunk 2]
```diff
@@ -1504,6 +1504,9 @@ type RuntimeConfig struct {
 	//
 	Watches []map[string]interface{}
 
+	// EnqueueLimit is used to specify a timeout value of Raft command apply
+	EnqueueLimit time.Duration
```
Comment (i0rek) [inline agent/config/runtime.go]: Your comment says timeout and thats also the name for this variable in raft: https://github.com/hashicorp/raft/blob/e4251bbbec4d7ad67b394786de0c0d908d8d6253/api.go#L674. We have a couple of raft related configuration options and I would like to suggest another name: `RaftApplyTimeout`. Let me know if you have another idea.

I know this name existed before, but lets take that opportunity and make this piece of code a better place :).

File: agent/config/runtime_test.go [hunk 3]
```diff
@@ -6018,6 +6018,7 @@ func TestSanitize(t *testing.T) {
 		"EnterpriseRuntimeConfig": ` + entRuntimeConfigSanitize + `,
 		"ExposeMaxPort": 0,
 		"ExposeMinPort": 0,
+		"EnqueueLimit": "0s",
```
Comment (i0rek) [inline agent/config/runtime_test.go]: Please check other config options, there are more tests in this file for `json` and `hcl` representations.

File: agent/config/config.go (hunk 1 above)
Comment (mitake) [reply agent/config/config.go]: Thanks, I moved the place in the struct.

File: agent/config/runtime.go (hunk 2 above)
Comment (mitake) [reply agent/config/runtime.go]: I see. I renamed the variable names to `RaftApplyTimeout`.

File: agent/config/runtime_test.go (hunk 3 above)
Comment (mitake) [reply agent/config/runtime_test.go]: Ah I see. I also updated `TestFullConfig`. Do I have other places to update? Thanks for catching this!

**Comment by mitake** [comment]
@i0rek thanks for your review. I updated on your comments. But on my local env `make test` failed with unclear reason like below. It also happens on the latest master. IIUC it didn't happen in the older version of the branch. Do you know how to resolve it?
```
~/g/s/g/h/consul [enqueue-limit]× » make test                                                                                          00:45:50
--> Checking for other consul instances
==> Building Consul - OSes: linux, Architectures: amd64
Building sequentially with go install
--->   linux/amd64
--> Running go vet
go test -tags '' -i ./... ./api/... ./sdk/...
go: directory api is outside main module
go: directory api/watch is outside main module
go: directory sdk/freeport is outside main module
go: directory sdk/testutil is outside main module
go: directory sdk/testutil/retry is outside main module
GNUmakefile:220: recipe for target 'test-install-deps' failed
make: *** [test-install-deps] Error 1
```

File: agent/config/runtime_test.go [hunk 4]
```diff
@@ -4283,7 +4283,8 @@ func TestFullConfig(t *testing.T) {
 					"key": "sl3Dffu7",
 					"args": ["dltjDJ2a", "flEa7C2d"]
 				}
-			]
+			],
+			"raft_apply_timeout": "0s"
```
Comment (i0rek) [inline agent/config/runtime_test.go]: In this test we are using some random values to make sure they come out properly. The problem with `0s` is that this is the default and we cannot tell if the conversion is working or not. This line and the next one in the hcl config should have the same random duration. eg. `20m44s`

Comment (mitake) [reply agent/config/runtime_test.go]: Ah I see. Thanks for pointing out! I fixed it in the latest branch. Could you take a look?

File: agent/agent.go [hunk 5]
```diff
@@ -1396,6 +1396,8 @@ func (a *Agent) consulConfig() (*consul.Config, error) {
 
 	base.ConfigEntryBootstrap = a.config.ConfigEntryBootstrap
 
+	base.RaftApplyTimeout = a.config.RaftApplyTimeout
```
Comment (i0rek) [inline agent/agent.go]: I just noticed that it has to be checked if the Timeout is `0` because otherwise that would override the default.

**Comment by i0rek** [comment]
There is another way we could deal with it, by adding it here: https://github.com/indeedeng/consul/blob/89def2e0170dbebb71e793aca72820bdac526ecd/agent/config/default.go#L30. One way or another, the `0` shouldn't override the default.

File: agent/agent.go (hunk 5 above)
Comment (mitake) [reply agent/agent.go]: Ahh thanks for pointing out! I'll update `DefaultSource()` for this problem.

Comment (mitake) [reply agent/agent.go]: Resolved in the latest master, could you take a look?

File: agent/consul/rpc.go [hunk 6]
```diff
@@ -37,11 +37,11 @@ const (
 	// If it's over 1MB something is probably being abusive.
 	raftWarnSize = 1024 * 1024
 
-	// enqueueLimit caps how long we will wait to enqueue
+	// defaultRaftApplyTimeout caps how long we will wait to enqueue
 	// a new Raft command. Something is probably wrong if this
 	// value is ever reached. However, it prevents us from blocking
 	// the requesting goroutine forever.
-	enqueueLimit = 30 * time.Second
+	defaultRaftApplyTimeout = 30 * time.Second
```
Comment (i0rek) [inline agent/consul/rpc.go]: Since the default from agent/config/default.go is no being used, I don't think we need this one anymore.

Comment (mitake) [reply agent/consul/rpc.go]: ah I missed the overwriting in DefaultConfig(), thanks for pointing.

**Comment by mitake** [comment]
@i0rek updated based on your comment, could you take a look?

**Comment by i0rek** [comment]
CI didn't test because some files didn't get formatted with `go fmt`. If you check the details in Circle, it will tell you.

**Comment by mitake** [comment]
@i0rek sorry for that, I formatted the file and triggered CI again. Probably the failed test would be a freaky one?

**Comment by i0rek** [comment]
@mitake I was just wondering under which circumstances you are reaching the hard coded timeout?

**Comment by mitake** [comment]
@i0rek Got it, I'll update the document some time (hopefully) this week. But if it's not costly for you, can I ask you to update the doc? It's really helpful for me :)

We have multiple extremely stressed consul clusters and the problem happens occasionally. I'll talk with other people in the company about making some of our metrics public (not sure it's ok though).

**Comment by mitake** [comment]
https://github.com/hashicorp/consul/pull/7237 This PR shares the motivation of handling heavy traffic. I'm glad if I can have your comments on it :)

**Comment by i0rek** [comment]
Thank you for providing more context. There is one thing that I am a little bit worried about: adding more knobs and dials to Consul comes at a cost - we have to maintain it and it makes operating Consul harder in a way because there are more things to tune.

I noticed this a little late in the process of this PR, but I cannot imagine in which cases you would want to increase this timeout. If it takes > 30 seconds already, something is probably very wrong. You probably shouldn't increase this timeout, but instead fix your performance problem.

Right now I am not convinced that this value should be increased, because it already is pretty high. If you could share more of your architecture and what you are struggling with exactly, that would help narrowing down the root cause.

**Comment by mitake** [comment]
@dnephin thanks for taking a look, I understood. It's great if I can get your opinions on https://github.com/hashicorp/consul/pull/7237 :)

**Pull request closed without merging**